File: scripts/validate_deals.py
#!/usr/bin/env python3
"""
Validates student-benefits-2026.ts for:
1. Duplicate slugs
2. Duplicate companies (different slugs for same company)
3. Broken/dead links (HTTP status checks)
4. Missing required fields
5. Placeholder/empty values
6. Invalid URLs
7. Duplicate titles
"""

import json
import re
import sys
import subprocess
import concurrent.futures
import urllib.request
import urllib.error
import http.client
from pathlib import Path

DATA_FILE = Path('/Users/raviteja/KIRO/FoundersPrime/data/student-benefits-2026.ts')
REPORT_FILE = Path('/tmp/validation_report.json')

# Required fields
REQUIRED_FIELDS = ['title', 'company', 'slug', 'category', 'appCategory', 
                   'offerSummary', 'description', 'benefitType', 'value',
                   'eligibility', 'region', 'verification', 'url', 'claimUrl']

VALID_BENEFIT_TYPES = {'Free', 'Discount', 'Credits', 'Grant', 'Scholarship', 'Program', 'Other'}
VALID_APP_CATEGORIES = {'Software & Tools', 'Credits & Savings', 'Funding & Opportunities', 
                        'Programs', 'Lifestyle'}
VALID_REGIONS = {'Global', 'US', 'India', 'UK', 'EU', 'Limited Countries', 'US & Canada',
                 'US, Canada', 'North America', 'Asia', 'Europe'}

def parse_ts_file(filepath):
    """Parse the TypeScript array of objects from the .ts file"""
    content = filepath.read_text(encoding='utf-8')
    
    # Extract the JSON array content between the first [ and last ]
    # The file has: export const studentBenefits2026: StudentBenefit[] = [...]
    match = re.search(r'=\s*(\[[\s\S]*\]);?\s*$', content)
    if not match:
        print("ERROR: Could not find array in file", file=sys.stderr)
        sys.exit(1)
    
    array_str = match.group(1)
    
    # Fix TypeScript-specific syntax for JSON parsing
    # Remove trailing commas before } and ]
    array_str = re.sub(r',(\s*[\]}])', r'\1', array_str)
    # Remove TypeScript comments
    array_str = re.sub(r'//[^\n]*', '', array_str)
    array_str = re.sub(r'/\*[\s\S]*?\*/', '', array_str)
    
    try:
        return json.loads(array_str)
    except json.JSONDecodeError as e:
        print(f"ERROR parsing JSON: {e}", file=sys.stderr)
        # Try to find the problematic line
        lines = array_str.split('\n')
        err_line = e.lineno
        print(f"Context around line {err_line}:", file=sys.stderr)
        for i in range(max(0, err_line-3), min(len(lines), err_line+3)):
            print(f"  {i+1}: {lines[i]}", file=sys.stderr)
        sys.exit(1)

def check_url(entry):
    """Check if a URL is reachable. Returns (slug, url, status, is_redirect, final_url)"""
    slug = entry.get('slug', 'unknown')
    url = entry.get('claimUrl') or entry.get('url', '')
    
    if not url or url.strip() == '':
        return slug, url, 'EMPTY', False, ''
    
    if not url.startswith('http'):
        return slug, url, 'INVALID_SCHEME', False, ''
    
    try:
        req = urllib.request.Request(url, method='HEAD', headers={
            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120 Safari/537.36',
            'Accept': 'text/html,application/xhtml+xml,*/*',
        })
        with urllib.request.urlopen(req, timeout=8) as resp:
            status = resp.status
            final_url = resp.url
            is_redirect = final_url != url
            return slug, url, str(status), is_redirect, final_url
    except urllib.error.HTTPError as e:
        return slug, url, f'HTTP_{e.code}', False, ''
    except urllib.error.URLError as e:
        reason = str(e.reason) if hasattr(e, 'reason') else str(e)
        return slug, url, f'ERROR: {reason[:60]}', False, ''
    except Exception as e:
        return slug, url, f'ERROR: {str(e)[:60]}', False, ''

def main():
    print("=" * 70)
    print("STUDENT BENEFITS DATA VALIDATION")
    print("=" * 70)
    
    print(f"\n📂 Loading data from {DATA_FILE.name}...")
    benefits = parse_ts_file(DATA_FILE)
    print(f"   ✓ Loaded {len(benefits)} entries\n")
    
    issues = {
        'duplicate_slugs': [],
        'duplicate_companies': [],
        'duplicate_titles': [],
        'missing_fields': [],
        'empty_values': [],
        'invalid_urls': [],
        'broken_links': [],
        'working_links': [],
        'redirects': [],
        'invalid_categories': [],
        'invalid_benefit_types': [],
        'suspicious_descriptions': [],
    }
    
    # ── 1. Check for duplicate slugs ─────────────────────────────────────
    print("🔍 Checking for duplicate slugs...")
    slug_map = {}
    for i, b in enumerate(benefits):
        slug = b.get('slug', '')
        if slug in slug_map:
            issues['duplicate_slugs'].append({
                'slug': slug,
                'first_idx': slug_map[slug],
                'second_idx': i,
                'companies': [benefits[slug_map[slug]].get('company'), b.get('company')]
            })
        else:
            slug_map[slug] = i
    print(f"   {'⚠️  ' + str(len(issues['duplicate_slugs'])) + ' duplicates found' if issues['duplicate_slugs'] else '✓ No duplicate slugs'}")
    
    # ── 2. Check for duplicate companies ─────────────────────────────────
    print("🔍 Checking for duplicate companies...")
    company_map = {}
    for i, b in enumerate(benefits):
        co = b.get('company', '').strip().lower()
        if co in company_map:
            # Only flag if title/offer differs (not just a second plan)
            existing = company_map[co]
            if b.get('value') != benefits[existing].get('value'):
                issues['duplicate_companies'].append({
                    'company': b.get('company'),
                    'idx1': existing,
                    'slug1': benefits[existing].get('slug'),
                    'value1': benefits[existing].get('value'),
                    'idx2': i,
                    'slug2': b.get('slug'),
                    'value2': b.get('value'),
                })
        else:
            company_map[co] = i
    print(f"   {'⚠️  ' + str(len(issues['duplicate_companies'])) + ' potential duplicate companies' if issues['duplicate_companies'] else '✓ No duplicate companies'}")
    
    # ── 3. Check for duplicate titles ────────────────────────────────────
    print("🔍 Checking for duplicate titles...")
    title_map = {}
    for i, b in enumerate(benefits):
        title = b.get('title', '').strip().lower()
        if title and title in title_map:
            issues['duplicate_titles'].append({
                'title': b.get('title'),
                'slugs': [benefits[title_map[title]].get('slug'), b.get('slug')]
            })
        else:
            title_map[title] = i
    print(f"   {'⚠️  ' + str(len(issues['duplicate_titles'])) + ' duplicate titles' if issues['duplicate_titles'] else '✓ No duplicate titles'}")
    
    # ── 4. Check missing required fields ─────────────────────────────────
    print("🔍 Checking for missing required fields...")
    for b in benefits:
        missing = [f for f in REQUIRED_FIELDS if not (b.get(f) or "").strip()]
        if missing:
            issues['missing_fields'].append({
                'slug': b.get('slug', 'UNKNOWN'),
                'company': b.get('company', 'UNKNOWN'),
                'missing': missing
            })
    print(f"   {'⚠️  ' + str(len(issues['missing_fields'])) + ' entries with missing fields' if issues['missing_fields'] else '✓ All required fields present'}")
    
    # ── 5. Check for suspicious/empty values ─────────────────────────────
    print("🔍 Checking for suspicious values...")
    placeholders = ['n/a', 'tbd', 'todo', 'placeholder', 'unknown', 'test', '???', 'example', 'lorem']
    for b in benefits:
        for field in ['offerSummary', 'description', 'value']:
            val = (b.get(field) or '').lower().strip()
            if any(p == val for p in placeholders):
                issues['empty_values'].append({
                    'slug': b.get('slug'),
                    'company': b.get('company'),
                    'field': field,
                    'value': b.get(field)
                })
    print(f"   {'⚠️  ' + str(len(issues['empty_values'])) + ' suspicious values' if issues['empty_values'] else '✓ No suspicious/placeholder values'}")
    
    # ── 6. Validate URL format ────────────────────────────────────────────
    print("🔍 Validating URL formats...")
    url_pattern = re.compile(r'^https?://[^\s/$.?#].[^\s]*$')
    for b in benefits:
        for field in ['url', 'claimUrl']:
            url = b.get(field, '')
            if url and not url_pattern.match(url):
                issues['invalid_urls'].append({
                    'slug': b.get('slug'),
                    'company': b.get('company'),
                    'field': field,
                    'url': url
                })
    print(f"   {'⚠️  ' + str(len(issues['invalid_urls'])) + ' malformed URLs' if issues['invalid_urls'] else '✓ All URLs have valid format'}")
    
    # ── 7. Check invalid categories/benefit types ─────────────────────────
    print("🔍 Checking field values...")
    for b in benefits:
        bt = b.get('benefitType', '')
        if bt and bt not in VALID_BENEFIT_TYPES:
            issues['invalid_benefit_types'].append({
                'slug': b.get('slug'),
                'company': b.get('company'),
                'benefitType': bt
            })
        ac = b.get('appCategory', '')
        if ac and ac not in VALID_APP_CATEGORIES:
            issues['invalid_categories'].append({
                'slug': b.get('slug'),
                'company': b.get('company'),
                'appCategory': ac
            })
    
    total_category_issues = len(issues['invalid_categories']) + len(issues['invalid_benefit_types'])
    print(f"   {'⚠️  ' + str(total_category_issues) + ' invalid category/type values' if total_category_issues else '✓ All categories and benefit types are valid'}")
    
    # ── 8. Link checking (sample 200 unique URLs) ─────────────────────────
    print("\n🔗 Checking links (sampling up to 200 unique claimUrls)...")
    seen_urls = set()
    entries_to_check = []
    for b in benefits:
        url = b.get('claimUrl') or b.get('url', '')
        if url and url not in seen_urls:
            seen_urls.add(url)
            entries_to_check.append(b)
            if len(entries_to_check) >= 200:
                break
    
    print(f"   Checking {len(entries_to_check)} unique URLs (parallel, 30 threads)...")
    
    with concurrent.futures.ThreadPoolExecutor(max_workers=30) as executor:
        futures = {executor.submit(check_url, e): e for e in entries_to_check}
        done_count = 0
        for future in concurrent.futures.as_completed(futures):
            slug, url, status, is_redirect, final_url = future.result()
            done_count += 1
            if done_count % 50 == 0:
                print(f"   ... {done_count}/{len(entries_to_check)} checked")
            
            if status.startswith('2') or status.startswith('3'):
                issues['working_links'].append({'slug': slug, 'url': url, 'status': status})
                if is_redirect and status.startswith('3'):
                    issues['redirects'].append({'slug': slug, 'from': url, 'to': final_url, 'status': status})
            elif status == 'EMPTY':
                issues['broken_links'].append({'slug': slug, 'url': url, 'status': 'EMPTY_URL', 'severity': 'HIGH'})
            elif 'HTTP_404' in status or 'HTTP_410' in status:
                issues['broken_links'].append({'slug': slug, 'url': url, 'status': status, 'severity': 'HIGH'})
            elif 'HTTP_5' in status:
                issues['broken_links'].append({'slug': slug, 'url': url, 'status': status, 'severity': 'MEDIUM'})
            elif 'ERROR' in status or 'INVALID' in status:
                issues['broken_links'].append({'slug': slug, 'url': url, 'status': status, 'severity': 'HIGH'})
            # HTTP 401, 403, 405 etc. may just block bots — mark as WARNING not broken
            elif 'HTTP_40' in status:
                issues['redirects'].append({'slug': slug, 'url': url, 'status': status + ' (bot-blocked, likely fine)'})
    
    print(f"\n   ✅ Working: {len(issues['working_links'])} links")
    print(f"   ⚠️  Redirects: {len(issues['redirects'])}")
    print(f"   ❌ Broken: {len(issues['broken_links'])}")
    
    # ── 9. Write report ───────────────────────────────────────────────────
    report = {
        'summary': {
            'total_entries': len(benefits),
            'duplicate_slugs': len(issues['duplicate_slugs']),
            'duplicate_companies': len(issues['duplicate_companies']),
            'duplicate_titles': len(issues['duplicate_titles']),
            'missing_fields': len(issues['missing_fields']),
            'empty_values': len(issues['empty_values']),
            'invalid_urls': len(issues['invalid_urls']),
            'invalid_categories': len(issues['invalid_categories']),
            'invalid_benefit_types': len(issues['invalid_benefit_types']),
            'broken_links': len(issues['broken_links']),
            'working_links': len(issues['working_links']),
            'redirects': len(issues['redirects']),
            'urls_checked': len(entries_to_check),
        },
        'issues': issues
    }
    
    REPORT_FILE.write_text(json.dumps(report, indent=2, ensure_ascii=False), encoding='utf-8')
    print(f"\n📄 Full report saved to {REPORT_FILE}")
    
    # ── Print summary ─────────────────────────────────────────────────────
    print("\n" + "=" * 70)
    print("SUMMARY")
    print("=" * 70)
    s = report['summary']
    print(f"Total entries:         {s['total_entries']}")
    print(f"Duplicate slugs:       {s['duplicate_slugs']} {'❌' if s['duplicate_slugs'] else '✅'}")
    print(f"Duplicate companies:   {s['duplicate_companies']} {'⚠️' if s['duplicate_companies'] else '✅'}")
    print(f"Duplicate titles:      {s['duplicate_titles']} {'⚠️' if s['duplicate_titles'] else '✅'}")
    print(f"Missing fields:        {s['missing_fields']} {'❌' if s['missing_fields'] else '✅'}")
    print(f"Suspicious values:     {s['empty_values']} {'⚠️' if s['empty_values'] else '✅'}")
    print(f"Malformed URLs:        {s['invalid_urls']} {'❌' if s['invalid_urls'] else '✅'}")
    print(f"Invalid categories:    {s['invalid_categories']} {'⚠️' if s['invalid_categories'] else '✅'}")
    print(f"Invalid benefit types: {s['invalid_benefit_types']} {'⚠️' if s['invalid_benefit_types'] else '✅'}")
    print(f"\nLink Check ({s['urls_checked']} URLs sampled):")
    print(f"  Working:   {s['working_links']} ✅")
    print(f"  Redirects: {s['redirects']} ⚠️")
    print(f"  Broken:    {s['broken_links']} {'❌' if s['broken_links'] else '✅'}")
    
    # Print broken links details
    if issues['broken_links']:
        print("\n❌ BROKEN LINKS:")
        for bl in issues['broken_links']:
            print(f"  [{bl['severity']}] {bl['slug']}: {bl['status']}")
            print(f"         {bl['url']}")
    
    # Print duplicate slugs
    if issues['duplicate_slugs']:
        print("\n❌ DUPLICATE SLUGS:")
        for d in issues['duplicate_slugs']:
            print(f"  Slug '{d['slug']}': entries {d['first_idx']} ({d['companies'][0]}) and {d['second_idx']} ({d['companies'][1]})")
    
    # Print duplicate companies (top 10)
    if issues['duplicate_companies']:
        print(f"\n⚠️  DUPLICATE COMPANIES (showing first 10 of {len(issues['duplicate_companies'])}):")
        for d in issues['duplicate_companies'][:10]:
            print(f"  {d['company']}: '{d['value1']}' (slug: {d['slug1']}) vs '{d['value2']}' (slug: {d['slug2']})")
    
    return report

if __name__ == '__main__':
    main()

File: archive/scripts/validate_deals.js
#!/usr/bin/env node
/**
 * Validates student-benefits-2026.ts for:
 * 1. Duplicate slugs
 * 2. Duplicate companies
 * 3. Missing required fields
 * 4. Empty/placeholder values
 * 5. Malformed URLs
 * 6. Broken links (HTTP checks, sampled)
 */

import { readFileSync, writeFileSync } from 'fs';
import https from 'https';
import http from 'http';

const DATA_FILE = 'data/student-benefits-2026.ts';
const REPORT_FILE = '/tmp/validation_report.json';

const REQUIRED_FIELDS = ['title', 'company', 'slug', 'category', 'appCategory',
  'offerSummary', 'description', 'benefitType', 'value',
  'eligibility', 'region', 'verification', 'url', 'claimUrl'];

const VALID_BENEFIT_TYPES = new Set(['Free', 'Discount', 'Credits', 'Grant', 'Scholarship', 'Program', 'Other']);
const VALID_APP_CATEGORIES = new Set(['Software & Tools', 'Credits & Savings', 'Funding & Opportunities', 'Programs', 'Lifestyle']);
const PLACEHOLDERS = new Set(['n/a', 'tbd', 'todo', 'placeholder', 'unknown', 'test', '???', 'example', 'lorem ipsum', 'lorem']);

// ── Load data ──────────────────────────────────────────────────────────
console.log('='.repeat(70));
console.log('STUDENT BENEFITS DATA VALIDATION');
console.log('='.repeat(70));

const content = readFileSync(DATA_FILE, 'utf-8');
const match = content.match(/const studentBenefits2026\s*[=:][^=]*=\s*(\[[\s\S]*\]);/);
if (!match) {
  console.error('ERROR: Could not find array in file');
  process.exit(1);
}

const benefits = eval('(' + match[1] + ')');
console.log(`\n📂 Loaded ${benefits.length} entries from ${DATA_FILE}\n`);

const issues = {
  duplicate_slugs: [],
  duplicate_companies: [],
  duplicate_titles: [],
  missing_fields: [],
  empty_values: [],
  invalid_urls: [],
  invalid_categories: [],
  invalid_benefit_types: [],
  broken_links: [],
  working_links: [],
  bot_blocked: [],
  redirects: [],
};

// ── 1. Duplicate slugs ─────────────────────────────────────────────────
console.log('🔍 Checking for duplicate slugs...');
const slugMap = new Map();
benefits.forEach((b, i) => {
  const slug = b.slug || '';
  if (slugMap.has(slug)) {
    issues.duplicate_slugs.push({
      slug,
      first_idx: slugMap.get(slug),
      second_idx: i,
      companies: [benefits[slugMap.get(slug)].company, b.company],
    });
  } else {
    slugMap.set(slug, i);
  }
});
console.log(`   ${issues.duplicate_slugs.length ? '⚠️  ' + issues.duplicate_slugs.length + ' duplicates' : '✓ None'}`);

// ── 2. Duplicate companies ─────────────────────────────────────────────
console.log('🔍 Checking for duplicate companies...');
const companyMap = new Map();
benefits.forEach((b, i) => {
  const co = (b.company || '').toLowerCase().trim();
  if (companyMap.has(co)) {
    const existingIdx = companyMap.get(co);
    const existing = benefits[existingIdx];
    // Flag if they have different values/offers (different plans are OK)
    if (b.value !== existing.value) {
      issues.duplicate_companies.push({
        company: b.company,
        slug1: existing.slug, value1: existing.value,
        slug2: b.slug, value2: b.value,
      });
    }
  } else {
    companyMap.set(co, i);
  }
});
console.log(`   ${issues.duplicate_companies.length ? '⚠️  ' + issues.duplicate_companies.length + ' duplicate companies' : '✓ None'}`);

// ── 3. Duplicate titles ───────────────────────────────────────────────
console.log('🔍 Checking for duplicate titles...');
const titleMap = new Map();
benefits.forEach((b, i) => {
  const title = (b.title || '').toLowerCase().trim();
  if (titleMap.has(title)) {
    issues.duplicate_titles.push({
      title: b.title,
      slug1: benefits[titleMap.get(title)].slug,
      slug2: b.slug,
    });
  } else {
    titleMap.set(title, i);
  }
});
console.log(`   ${issues.duplicate_titles.length ? '⚠️  ' + issues.duplicate_titles.length + ' duplicate titles' : '✓ None'}`);

// ── 4. Missing required fields ─────────────────────────────────────────
console.log('🔍 Checking for missing required fields...');
benefits.forEach(b => {
  const missing = REQUIRED_FIELDS.filter(f => !(b[f] || '').toString().trim());
  if (missing.length) {
    issues.missing_fields.push({ slug: b.slug || 'UNKNOWN', company: b.company || 'UNKNOWN', missing });
  }
});
console.log(`   ${issues.missing_fields.length ? '⚠️  ' + issues.missing_fields.length + ' entries with missing fields' : '✓ All present'}`);

// ── 5. Suspicious/placeholder values ─────────────────────────────────
console.log('🔍 Checking for placeholder/suspicious values...');
benefits.forEach(b => {
  ['offerSummary', 'description', 'value'].forEach(field => {
    const val = (b[field] || '').toLowerCase().trim();
    if (PLACEHOLDERS.has(val) || val.length < 5) {
      issues.empty_values.push({ slug: b.slug, company: b.company, field, value: b[field] });
    }
  });
});
console.log(`   ${issues.empty_values.length ? '⚠️  ' + issues.empty_values.length + ' suspicious values' : '✓ None'}`);

// ── 6. URL format validation ──────────────────────────────────────────
console.log('🔍 Validating URL formats...');
const urlPattern = /^https?:\/\/[^\s/$.?#].[^\s]*$/;
benefits.forEach(b => {
  ['url', 'claimUrl'].forEach(field => {
    const url = b[field] || '';
    if (url && !urlPattern.test(url)) {
      issues.invalid_urls.push({ slug: b.slug, company: b.company, field, url });
    }
  });
});
console.log(`   ${issues.invalid_urls.length ? '⚠️  ' + issues.invalid_urls.length + ' malformed URLs' : '✓ All valid format'}`);

// ── 7. Category/type validation ───────────────────────────────────────
console.log('🔍 Checking categories and benefit types...');
benefits.forEach(b => {
  if (b.benefitType && !VALID_BENEFIT_TYPES.has(b.benefitType)) {
    issues.invalid_benefit_types.push({ slug: b.slug, company: b.company, benefitType: b.benefitType });
  }
  if (b.appCategory && !VALID_APP_CATEGORIES.has(b.appCategory)) {
    issues.invalid_categories.push({ slug: b.slug, company: b.company, appCategory: b.appCategory });
  }
});
const catIssues = issues.invalid_categories.length + issues.invalid_benefit_types.length;
console.log(`   ${catIssues ? '⚠️  ' + catIssues + ' invalid category/type values' : '✓ All valid'}`);

// ── 8. HTTP link checking ─────────────────────────────────────────────
console.log('\n🔗 Sampling links for HTTP status checks...');

function checkUrl(entry) {
  return new Promise((resolve) => {
    const url = entry.claimUrl || entry.url || '';
    const slug = entry.slug || 'unknown';
    
    if (!url || !url.startsWith('http')) {
      return resolve({ slug, url, status: 'EMPTY_OR_INVALID', severity: 'HIGH' });
    }
    
    let urlObj;
    try { urlObj = new URL(url); } catch(e) {
      return resolve({ slug, url, status: 'MALFORMED_URL', severity: 'HIGH' });
    }
    
    const lib = url.startsWith('https') ? https : http;
    const opts = {
      method: 'HEAD',
      hostname: urlObj.hostname,
      path: urlObj.pathname + urlObj.search,
      port: urlObj.port || (url.startsWith('https') ? 443 : 80),
      timeout: 8000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 Chrome/120 Safari/537.36',
        'Accept': 'text/html,*/*',
      },
    };
    
    const req = lib.request(opts, (res) => {
      const status = res.statusCode;
      const location = res.headers.location || '';
      if (status >= 200 && status < 400) {
        resolve({ slug, url, status, location, ok: true });
      } else if (status === 401 || status === 403 || status === 405 || status === 429) {
        resolve({ slug, url, status, ok: true, note: 'bot-blocked' });
      } else if (status === 404 || status === 410 || status === 451) {
        resolve({ slug, url, status, severity: 'HIGH', ok: false });
      } else if (status >= 500) {
        resolve({ slug, url, status, severity: 'MEDIUM', ok: false });
      } else {
        resolve({ slug, url, status, ok: true });
      }
    });
    
    req.on('timeout', () => { req.destroy(); resolve({ slug, url, status: 'TIMEOUT', severity: 'LOW', ok: false }); });
    req.on('error', (e) => { resolve({ slug, url, status: 'NETWORK_ERROR: ' + e.message.slice(0,50), severity: 'MEDIUM', ok: false }); });
    req.end();
  });
}

// Sample: unique URLs, prioritize claimUrls that point to external sites
const seenUrls = new Set();
const toCheck = [];
for (const b of benefits) {
  const url = b.claimUrl || b.url || '';
  if (url && !seenUrls.has(url) && toCheck.length < 300) {
    seenUrls.add(url);
    toCheck.push(b);
  }
}
console.log(`   Checking ${toCheck.length} unique URLs (parallel batches of 40)...`);

async function checkBatch(batch) {
  return Promise.all(batch.map(checkUrl));
}

async function runAllChecks() {
  const batchSize = 40;
  const allResults = [];
  
  for (let i = 0; i < toCheck.length; i += batchSize) {
    const batch = toCheck.slice(i, i + batchSize);
    const results = await checkBatch(batch);
    allResults.push(...results);
    process.stdout.write(`   ... ${Math.min(i + batchSize, toCheck.length)}/${toCheck.length} checked\r`);
  }
  console.log(`   ✓ All ${toCheck.length} URLs checked             `);
  
  for (const r of allResults) {
    if (r.ok === true) {
      if (r.note === 'bot-blocked') {
        issues.bot_blocked.push(r);
      } else if (r.location) {
        issues.redirects.push(r);
        issues.working_links.push(r);
      } else {
        issues.working_links.push(r);
      }
    } else {
      issues.broken_links.push(r);
    }
  }
  
  console.log(`   ✅ Working: ${issues.working_links.length}`);
  console.log(`   🔄 Redirects: ${issues.redirects.length}`);
  console.log(`   🤖 Bot-blocked (likely fine): ${issues.bot_blocked.length}`);
  console.log(`   ❌ Broken: ${issues.broken_links.length}`);
  
  // ── Write report ─────────────────────────────────────────────────────
  const report = {
    summary: {
      total_entries: benefits.length,
      duplicate_slugs: issues.duplicate_slugs.length,
      duplicate_companies: issues.duplicate_companies.length,
      duplicate_titles: issues.duplicate_titles.length,
      missing_fields: issues.missing_fields.length,
      empty_values: issues.empty_values.length,
      invalid_urls: issues.invalid_urls.length,
      invalid_categories: issues.invalid_categories.length,
      invalid_benefit_types: issues.invalid_benefit_types.length,
      broken_links: issues.broken_links.length,
      working_links: issues.working_links.length,
      bot_blocked: issues.bot_blocked.length,
      redirects: issues.redirects.length,
      urls_checked: toCheck.length,
    },
    issues,
  };
  
  writeFileSync(REPORT_FILE, JSON.stringify(report, null, 2), 'utf-8');
  
  // ── Print summary ─────────────────────────────────────────────────────
  const s = report.summary;
  console.log('\n' + '='.repeat(70));
  console.log('FINAL SUMMARY');
  console.log('='.repeat(70));
  console.log(`Total entries:         ${s.total_entries}`);
  console.log(`Duplicate slugs:       ${s.duplicate_slugs}  ${s.duplicate_slugs ? '❌' : '✅'}`);
  console.log(`Duplicate companies:   ${s.duplicate_companies}  ${s.duplicate_companies ? '⚠️' : '✅'}`);
  console.log(`Duplicate titles:      ${s.duplicate_titles}  ${s.duplicate_titles ? '⚠️' : '✅'}`);
  console.log(`Missing fields:        ${s.missing_fields}  ${s.missing_fields ? '⚠️' : '✅'}`);
  console.log(`Suspicious values:     ${s.empty_values}  ${s.empty_values ? '⚠️' : '✅'}`);
  console.log(`Malformed URLs:        ${s.invalid_urls}  ${s.invalid_urls ? '❌' : '✅'}`);
  console.log(`Invalid categories:    ${s.invalid_categories}  ${s.invalid_categories ? '⚠️' : '✅'}`);
  console.log(`Invalid benefit types: ${s.invalid_benefit_types}  ${s.invalid_benefit_types ? '⚠️' : '✅'}`);
  console.log(`\nLink Check (${s.urls_checked} URLs sampled):`);
  console.log(`  ✅ Working:       ${s.working_links}`);
  console.log(`  🔄 Redirects:     ${s.redirects}`);
  console.log(`  🤖 Bot-blocked:   ${s.bot_blocked}`);
  console.log(`  ❌ Broken:        ${s.broken_links}`);
  
  if (issues.duplicate_slugs.length) {
    console.log('\n❌ DUPLICATE SLUGS:');
    issues.duplicate_slugs.forEach(d => {
      console.log(`  Slug '${d.slug}': ${d.companies[0]} (idx ${d.first_idx}) vs ${d.companies[1]} (idx ${d.second_idx})`);
    });
  }
  
  if (issues.broken_links.length) {
    console.log('\n❌ BROKEN LINKS:');
    issues.broken_links.forEach(b => {
      console.log(`  [${b.severity || 'MED'}] ${b.slug}: HTTP ${b.status}`);
      console.log(`        ${b.url}`);
    });
  }
  
  if (issues.duplicate_companies.length) {
    console.log(`\n⚠️  DUPLICATE COMPANIES (top 15 of ${issues.duplicate_companies.length}):`);
    issues.duplicate_companies.slice(0, 15).forEach(d => {
      console.log(`  ${d.company}: [${d.slug1}] "${d.value1}"  vs  [${d.slug2}] "${d.value2}"`);
    });
  }
  
  if (issues.invalid_categories.length) {
    console.log('\n⚠️  INVALID APP CATEGORIES:');
    issues.invalid_categories.slice(0, 10).forEach(c => {
      console.log(`  ${c.slug}: "${c.appCategory}"`);
    });
  }
  
  if (issues.missing_fields.length) {
    console.log('\n⚠️  MISSING FIELDS (top 10):');
    issues.missing_fields.slice(0, 10).forEach(m => {
      console.log(`  ${m.slug} (${m.company}): missing [${m.missing.join(', ')}]`);
    });
  }
  
  console.log(`\n📄 Full report: ${REPORT_FILE}`);
}

runAllChecks().catch(console.error);
